node/src/components/deploy_buffer.rs

```diff
@@ -332,7 +333,48 @@ impl DeployBuffer {
         let mut holds = HashSet::new();
         let mut have_hit_transfer_limit = false;
         let mut have_hit_deploy_limit = false;
-        for (with_approvals, footprint) in self.proposable() {
+
+        let proposable = self.proposable();
+
+        let mut buckets: HashMap<Digest, Vec<(DeployHashWithApprovals, DeployFootprint)>> =
+            HashMap::new();
+
+        for (with_approvals, footprint) in proposable {
+            let body_hash = *footprint.header.body_hash();
+            buckets
+                .entry(body_hash)
+                .and_modify(|vec| vec.push((with_approvals.clone(), footprint.clone())))
+                .or_insert(vec![(with_approvals, footprint)]);
+        }
+
+        let indexer = buckets.keys().cloned().collect_vec();
+        let mut idx = 0;
+
+        loop {
+            if buckets.is_empty() {
+                break;
+            }
+            let body_hash = match indexer.get(idx) {
+                None => {
+                    idx = 0; // reset outer loop
+                    continue;
+                }
+                Some(body_hash) => {
+                    idx += 1;
+                    body_hash
+                }
+            };
+            let (with_approvals, footprint) = match buckets.entry(*body_hash) {
+                Entry::Vacant(_) => continue,
+                Entry::Occupied(entry) => {
+                    let deploys = entry.into_mut();
+                    if deploys.is_empty() {
+                        buckets.remove(body_hash);
+                        continue;
+                    }
+                    deploys.remove(0)
```

Comment:
This is O(n), making the entire algorithm O(n^2) in theory in the case of having a large queue all under a single body hash. In reality, it will be bound by the deploys-per-block limit, so it'll be O(50n).

Either we should use a `VecDeque` instead of a `Vec`, or, if we can afford it, change the order by calling `deploys.pop()` instead.

Reply:
tests added, impl'd some of what you suggested